Use memcached flags to detect need for unmarshalling and decompressing

dylanahsmith opened the pull request:
## Problem

I would like a cleaner and more flexible ActiveSupport::Cache::MemcachedStore so we don't need to inherit from and patch its methods to change the serializer or add compression.  I think we should also support conditional compression based a minimum and maximum size, which requires marking the key as being compressed to detect whether it needs to be decompressed.

## Solution

memcached supports having flags associated with keys, which are typically used to indicate the format of the cache value.  The memcached gem allows these to be leveraged by providing a codec.  Dalli also uses these flags for serialization and conditional compression, so I've used the same serialized and compressed flags.

This PR doesn't add conditional compression, but that can be done in a follow-up PR.

maximebedard commented on lib/active_support/cache/memcached_snappy_store.rb:
```diff
@@ -12,6 +12,16 @@ module Cache
+        def self.decompress(source)
+          Snappy.inflate(source)
+        end
+      end
```
WDYT of keeping the interface to `#inflate` & `#deflate` since Zlib also follow the same interface?

dylanahsmith replied:
I had considered that.  The main advantage of this interface is that it is compatible with being re-used with Dalli using its `:compressor` option.  Additionally, this is the interface that `ActiveSupport::Gzip` implements.

maximebedard replied:
Ah TIL, sounds good to me !

maximebedard commented:
Do you think it's worth merging `compress: false` in the options as a default, otherwise this would compress it twice for no good reason and seems very error prone?

dylanahsmith commented:
> Do you think it's worth merging compress: false in the options as a default, otherwise this would compress it twice for no good reason and seems very error prone?

That makes sense.  It is an existing bug, but seems like an easy change to make in MemcachedSnappyStore.

Another fix I should port from our application is that `deserialize_entry` should pass `compress: false` to `Entry.new` so that it doesn't needlessly compress the value such that it needs to be decompressed when `Entry#value` is called.

maximebedard commented:
LGTM, but there seems to have a few missing tests around the legacy codepaths. E.g. commenting https://github.com/Shopify/memcached_store/pull/51/files#diff-c93d4bdbaeb83a34a57b7a442c4d2b5fR34, https://github.com/Shopify/memcached_store/pull/51/files#diff-c93d4bdbaeb83a34a57b7a442c4d2b5fR43 and https://github.com/Shopify/memcached_store/pull/51/files#diff-c93d4bdbaeb83a34a57b7a442c4d2b5fR55, should make some test fail. WDYT?

dylanahsmith commented:
Added tests for a couple of those code paths.  I removed the flag masking since we always pass 0 for the flags on store operations, so that is the input to the codec's encode method.
